src/HttpServer.php

```diff
@@ -60,48 +63,81 @@ public function addMiddleware($middleware)
-            if (!$headerCompleted) {
-                $headerDecoder->on('data', function ($header) use (&$request, &$data, &$headerCompleted) {
-                    $request = RingCentral\Psr7\parse_request($header);
-                    $data = substr($data, strlen($header));
-                    $headerCompleted = true;
-                });
-                $headerStream->emit('data', array($data));
-            }
+                if ($request->hasHeader('Content-Length')) {
+                    $contentLength = $request->getHeaderLine('Content-Length');
 
-            if (isset($request)) {
-                if ($that->isChunkedEncodingActive($request)) {
-                    $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $that, $connection) {
-                        $bodyBuffer .= $chunk;
-                        if (strlen($chunk) == 0) {
-                            $that->sendBody($bodyBuffer, $connection, $request);
-                        }
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
+                    $int = (int) $contentLength;
+                    if ((string)$int !== (string)$contentLength) {
+                        // Send 400 status code if the value of 'Content-Length'
+                        // is not an integer or is duplicated
+                        $that->sendResponse(new Response(400), $connection);
+                        return;
                     }
                 }
+
+                $that->handleBody($request, $connection);
+
+                // remove header from $data, only body is left
+                $data = (string)substr($data, strlen($fullHeader));
+                if ($data !== '') {
+                    $connection->emit('data', array($data));
+                }
             }
-        });
+        };
+
+        $connection->on('data', $listener);
+    }
+
+    /** @internal */
+    public function handleBody(RequestInterface $request, ConnectionInterface $connection)
+    {
+        if ($this->isChunkedEncodingActive($request)) {
+            // Add ChunkedDecoder to stream
+            $chunkedDecoder = new ChunkedDecoder($connection);
+            $bodyStream = new HttpBodyStream($chunkedDecoder);
+            $request = $request->withBody($bodyStream);
+            $this->handleRequest($connection, $request);
+            return;
+        }
+
+        $contentLength = (int)$request->getHeaderLine('Content-Length');
```

Comment:
This is tested. The casting from the returned `null` from `getHeaderline` resulted in an 0 value. Therefore this scenario is ok. The lines below aren't needed. Will be removed in the next commit.